File: algorithms/cpp/3Sum Smaller.cpp
class Solution {
public:
    int threeSumSmaller(vector<int>& nums, int target) {
        if(nums.size() < 3) return 0;
        sort(nums.begin(), nums.end());
        int ret = 0;
        for(int i = 0;i < nums.size();++i) {
            int val = target - nums[i];
            
            int j = i + 1, k = nums.size() - 1;
            while(j < k) {
                if(nums[j] + nums[k] >= val) {
                    k--;
                }
                else { //if(nums[j] + nums[k] < val) {
                    ret += k - j;
                    j++;
                }
            }
        }

        return ret;
    }
};

File: algorithms/java/3Sum Smaller.java
public class Solution {
    public int threeSumSmaller(int[] nums, int target) {
        int n = nums.length;
        if(n < 3) {
            return 0;
        }
        Arrays.sort(nums);
        int ret = 0;
        for(int i = 0;i < n;++i) {
            int j = i + 1;
            int k = n - 1;
            while(j < k) {
                int t = nums[i] + nums[j] + nums[k];
                if(t < target) {
                    ret += k - j;
                    j++;
                } else if(t >= target) {
                    k--;
                }
            }
        }
        
        return ret;
    }
}